Add lockdown mode logging

karlcow opened the pull request:
#### 72d690ca074069922d8088c78ba5b2ab8fc9872e
<pre>
Add lockdown mode logging
<a href="https://bugs.webkit.org/show_bug.cgi?id=262518">https://bugs.webkit.org/show_bug.cgi?id=262518</a>
rdar://114657783

Reviewed by Chris Dumez.

When activated Lockdown mode is reducing some of the features
accessible to a Webpage. This creates breakages which can be
sometimes a bit hard to identify. The goal of this change is
to improve the findability of the type of issues by adding logging
for engineers to understand what type of breakage is going on.

Currently we only target fonts which have not been loaded. It seems,
for now, these are the main source of breakage. We initially wanted
to log the URL of the fonts being logged in syslog, but this create
significant issues in terms of privacy, as it would share URLs which
will be private to the user context, such as intranet systems. Logging
the name of the fonts might also create issues with inline nameless
fonts and we would only know the name given to the font at the CSS
parsing time, as the blocked fonts are not being downloaded.

This also adds logging in the Web Inspector console for fonts URLs
being blocked by Lockdown mode. Being private to the user context,
this is helpful for the Web developer to determine what fonts are
missing and gives the opportunity for them to share at their will
the issue in a bug report.

We might, in the future, add more engineering logging for other
features.

If you need to learn more about lockdown mode, see
<a href="https://support.apple.com/en-us/HT212650">https://support.apple.com/en-us/HT212650</a>

* Source/WebCore/loader/cache/AllowedFonts.cpp:
(WebCore::isFontBinaryAllowed):
* Source/WebCore/css/CSSFontFaceSrcValue.cpp:
(WebCore::CSSFontFaceSrcResourceValue::fontLoadRequest):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::fontLoadRequest):
* Source/WebCore/loader/cache/CachedFont.cpp:
(WebCore::CachedFont::finishLoading):
* Source/WebCore/loader/cache/CachedFont.h:
(WebCore::CachedFont::didRefuseToLoadCustomFont const):
* Source/WebCore/loader/cache/CachedFontLoadRequest.h:

Co-authored-by: Chris Dumez &lt;cdumez@apple.com&gt;
Co-authored-by: Brent Fulgham &lt;bfulgham@webkit.org&gt;
Canonical link: <a href="https://commits.webkit.org/269238@main">https://commits.webkit.org/269238@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/82beba7006691a1b55e6b3a9312c1ffb5c6ccfbe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21938 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22158 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23005 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23820 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20315 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22183 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26405 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22470 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21397 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22168 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21803 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19024 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24672 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/22118 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18931 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19878 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26143 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/19071 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19958 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20100 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24017 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/21313 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20536 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17496 "Found 1 new test failure: js/for-in-to-text.html (failure)") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/25356 "Built successfully") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/21749 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19888 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/19687 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5984 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5241 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24094 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/26629 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20487 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5816 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

karlcow commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -4585,8 +4586,12 @@ void WebPage::updatePreferences(const WebPreferencesStore& store)
         adjustSettingsForLockdownMode(settings, store);
+        // FIXME. Add the domain name?
+        // FIXME. Add RELEASE_LOG too?
+        m_page->console().addMessage(MessageSource::Security, MessageLevel::Info, makeString("This device has activated Lockdown Mode. It will affect some functionalities on the Web page."));
```
I'm not sure this is really working as intended yet. It needs to be double checked.

karlcow replied:
@brentfulgham guidance would be welcome here.

cdumez replied:
I doubt the console logging will be helpful here. My bet is that you'll want to move this check / logging to `WebLocalFrameLoaderClient::dispatchDidFinishLoad()` when `m_frame->isMainFrame()`. This way, it would lot once in the main frame after the load completes.

karlcow replied:
I removed it for now.

vitorroriz commented on Source/WebCore/loader/cache/AllowedFonts.cpp:
```diff
@@ -851,16 +852,20 @@ bool isFontBinaryAllowed(const void* data, size_t size, DownloadableBinaryFontAl
         auto sha = hashForFontData(data, size);
-        return allowedFontHashesInLockdownMode().contains(sha);
+        bool allowedFontHashes;
+        allowedFontHashes = allowedFontHashesInLockdownMode().contains(sha);
```
```suggestion
        auto allowedFontHashes = allowedFontHashesInLockdownMode().contains(sha);
```

karlcow replied:
thanks.

pxlcoder replied:
Doesn't look like this comment has been addressed yet, so marking as unresolved.

karlcow replied:
done on my local computer. :) but fair enough.

karlcow replied:
ok this time in the PR 😄

vitorroriz commented on Source/WebCore/loader/cache/AllowedFonts.cpp:
```diff
@@ -842,7 +843,7 @@ static AtomString hashForFontData(const void* data, size_t size)
 }
 
-bool isFontBinaryAllowed(const void* data, size_t size, DownloadableBinaryFontAllowedTypes allowedType)
+bool isFontBinaryAllowed(const void* data, size_t size, DownloadableBinaryFontAllowedTypes allowedType, std::string url)
```
url can be a `const String&` for avoiding copy, or a `StringView`

vitorroriz commented on Source/WebCore/loader/cache/AllowedFonts.cpp:
```diff
@@ -851,16 +852,20 @@ bool isFontBinaryAllowed(const void* data, size_t size, DownloadableBinaryFontAl
 }
 
-bool isFontBinaryAllowed(std::span<const uint8_t> data, DownloadableBinaryFontAllowedTypes allowedType)
+bool isFontBinaryAllowed(std::span<const uint8_t> data, DownloadableBinaryFontAllowedTypes allowedType, std::string url)
```
url can be a `const String&` for avoiding copy, or a `StringView`

vitorroriz commented on Source/WebCore/loader/cache/AllowedFonts.h:
```diff
@@ -41,7 +41,7 @@ enum class DownloadableBinaryFontAllowedTypes : uint8_t {
-bool isFontBinaryAllowed(std::span<const uint8_t>, DownloadableBinaryFontAllowedTypes);
-bool isFontBinaryAllowed(const void*, size_t, DownloadableBinaryFontAllowedTypes);
+bool isFontBinaryAllowed(std::span<const uint8_t>, DownloadableBinaryFontAllowedTypes, std::string url = "");
+bool isFontBinaryAllowed(const void*, size_t, DownloadableBinaryFontAllowedTypes, std::string url = "");
```
If you really need a default value for url, then instead of `const String&`, as I suggested in previous comments, it could be a `const String*` will `nullptr` as default. But make sure to check for null url.

cdumez reviewed (changes requested):
Please use `String` instead of `std::string`.

e.g.

`const String& url = emptyString()`

pxlcoder commented on Source/WebCore/loader/cache/AllowedFonts.cpp:
```diff
@@ -25,6 +25,7 @@
 
 #include "config.h"
 #include "AllowedFonts.h"
+#include "WebCore/css/FontFace.h"
```
```suggestion
#include "FontFace.h"
```

karlcow replied:
thanks. done!

pxlcoder commented on Source/WebCore/loader/cache/CachedFont.cpp:
```diff
@@ -41,6 +41,7 @@
 #include "TextResourceDecoder.h"
 #include "TypedElementDescendantIteratorInlines.h"
 #include "WOFFFileFormat.h"
+#include "WebCore/css/CSSFontFace.h"
```
```suggestion
#include "CSSFontFace.h"
```

karlcow replied:
thanks! done.

cdumez commented on Source/WebCore/loader/cache/AllowedFonts.cpp:
```diff
@@ -851,16 +852,19 @@ bool isFontBinaryAllowed(const void* data, size_t size, DownloadableBinaryFontAl
-        return allowedFontHashesInLockdownMode().contains(sha);
+        auto allowedFontHashes = allowedFontHashesInLockdownMode().contains(sha);
+        if (!allowedFontHashes)
+            RELEASE_LOG(Fonts, "[Lockdown Mode] This font is blocked: %s", url.utf8().data());
```
As far as I know, we don't allow logging URLs. It may be acceptable if you use:
`RELEASE_LOG(Fonts, "[Lockdown Mode] This font is blocked: %" SENSITIVE_LOG_STRING, url.utf8().data());`

Note that it will then not show in sysdiagnoses and you'll need enable sensitive logging on your machine when you want to see this logging. Otherwise, if you want to log in sysdiagnoses, I recommend logging something else than the URL (maybe font name?).

karlcow replied:
Ah! Good point. hmmm… 
And this is indeed the requested URL from the CSS. 

Probably we do not have access to the real font name at this stage (as the download is blocked because of lockdown mode), probably the parsed name from the CSS. 

Hmm… How do I access this?
But could this create also issues by revealing the name of the fonts in the log?

@brentfulgham what would be the more helpful choice, with regards to what you experienced in the past for Lockdown mode bugs?

brentfulgham replied:
I think just knowing that we blocked a font because of LDM is helpful. The best case would be to know the name of the font so we could add it to the set of fonts we review and approve.

However, since it will often be a potentially randomized data URL that may not be possible.

So for now, I think just noting the event is useful.

karlcow replied:
done.

cdumez commented on Source/WebCore/loader/cache/CachedFont.cpp:
```diff
@@ -74,7 +74,6 @@ bool CachedFont::shouldAllowCustomFont(const Ref<SharedBuffer>& data)
 {
     if (!m_loader || !m_loader->frame())
         return false;
-
```
Please omit this change from the PR.

karlcow replied:
done. thanks.

cdumez commented on Source/WebCore/loader/cache/AllowedFonts.cpp:
```diff
@@ -25,6 +25,7 @@
 
 #include "config.h"
 #include "AllowedFonts.h"
+#include "FontFace.h"
```
I doubt this new include is needed.

karlcow replied:
done. thanks.

cdumez reviewed (approved):
LGTM with nits.

pxlcoder reviewed (commented):
Are we going to pursue the console logging in a separate patch? If so, I think the commit message/description should be updated.

cdumez commented on Source/WebCore/loader/cache/AllowedFonts.cpp:
```diff
@@ -851,7 +851,10 @@ bool isFontBinaryAllowed(const void* data, size_t size, DownloadableBinaryFontAl
-        return allowedFontHashesInLockdownMode().contains(sha);
+        auto allowedFontHashes = allowedFontHashesInLockdownMode().contains(sha);
+        if (!allowedFontHashes)
+            RELEASE_LOG(Fonts, "[Lockdown Mode] A font with a forbidden type has been blocked.");
```
I think you'll need to include "Logging.h" to fix the build.

brentfulgham replied:
I'd like to see if you could extract the font file name from `data` (or perhaps pass it from the containing function). having the name of the font file would be immensely helpful to debugging.

karlcow replied:
@brentfulgham That seems problematic too, no? As if we do not log URI for privacy reasons, would it not be also an issue to log the name of the font face parsed from either the `src` or the `font-family` field .


```html
<style media="screen, print">
      @font-face {
        font-family: "MyPrivateFontOnIntranet";
        src: url("https://intranet.example.org/web-fonts/trackingCompanyInc.ttf");
      }
</style>
```

with cases also like these

```html
<style media="screen, print">
      @font-face {
        font-family: "MyPrivateFontOnIntranet";
        src: url("https://distributor.example.org/?id=345&version=italic");
      }
</style>
```

Or do we have access to the parsed font already. I though the current mechanism was blocking the download itself?

brentfulgham replied:
Actually, an even better approach would be:
1. Log to WTFLogAlways that some undisclosed font was blocked due to LDM.
2. Log to Web Inspector console that a font with a specific URL was blocked by LDM.

That way, when investigating LDM site failures we could see in the log that LDM blocked a font.

So your current patch is fine as-is, but I'd love a follw-up that told the WebInspector console that we had blocked a specific font. Then developers (or I) could look at that and figure out what was going on.

cdumez replied:
By `WTFLogAlways`, I assume you mean RELEASE_LOG() :)

I totally agree that we should log the failing URL when logging to the JS console. An earlier iteration was attempting to do this I believe though it was not working well. I had a proposal to fix it but Karl dropped it from the PR altogether.

karlcow replied:
As discussed with @brentfulgham we could try to log the url of the font directly in the console log. So at least for the web developers or the person doing the diagnosis there will be a chance to understand what is breaking without leaking it in sysdiagnose.

karlcow replied:
ok let me work on this today.

karlcow replied:
> An earlier iteration was attempting to do this I believe though it was not working well. I had a proposal to fix it but Karl dropped it from the PR altogether.

yes. The message alone, a font has been blocked didn't seem useful in console.log. But if we add the URL which stays private to the user context, then it's entirely a different game. But this will not work in AllowedFonts.cpp 

Chris [suggested](https://github.com/WebKit/WebKit/pull/18684#discussion_r1348178273)

[`WebLocalFrameLoaderClient::dispatchDidFinishLoad()`](https://searchfox.org/wubkat/rev/e21786e1635702a6ca7c1e30290f32e8a97fec23/Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp#723) when `m_frame->isMainFrame()`

status: digging in the code.

karlcow replied:
@cdumez hmm… dilemma…

[CachedFont::finishLoading](https://searchfox.org/wubkat/rev/e21786e1635702a6ca7c1e30290f32e8a97fec23/Source/WebCore/loader/cache/CachedFont.cpp#81-97) is the place where it is determined if the font should be loaded.

```cpp
if (!shouldAllowCustomFont(dataContiguous)) {
            setErrorAndDeleteData();
            return;
}
```

Above that, we do not know if the font will be blocked (lockdown mode or others). 
Is the console accessible at this stage?

cdumez replied:
It's a little bit of work but I think you can tweak this code here:
```
std::unique_ptr<FontLoadRequest> Document::fontLoadRequest(const String& url, bool isSVG, bool isInitiatingElementInUserAgentShadowTree, LoadedFromOpaqueSource loadedFromOpaqueSource)
{
    auto* cachedFont = m_fontLoader->cachedFont(completeURL(url), isSVG, isInitiatingElementInUserAgentShadowTree, loadedFromOpaqueSource);
    return cachedFont ? makeUnique<CachedFontLoadRequest>(*cachedFont) : nullptr;
}
```

to pass the Document to the CachedFontLoadRequest, like so:
```
std::unique_ptr<FontLoadRequest> Document::fontLoadRequest(const String& url, bool isSVG, bool isInitiatingElementInUserAgentShadowTree, LoadedFromOpaqueSource loadedFromOpaqueSource)
{
    auto* cachedFont = m_fontLoader->cachedFont(completeURL(url), isSVG, isInitiatingElementInUserAgentShadowTree, loadedFromOpaqueSource);
    return cachedFont ? makeUnique<CachedFontLoadRequest>(*cachedFont, *this) : nullptr;
}
```

Then you add a `Document&` parameter to the CachedFontLoadRequest constructor and save it as a data member:
```
WeakPtr<Document, WeakPtrImplWithEventTargetData> m_document;
```

Then in `CachedFont::finishLoading()`, set a boolean flag data member whenever we refuse to load the custom font:
e.g. `m_didRefuseToLoadCustomFont = true;`.

Also add a getter function for this flag.

Then in CachedFontLoadRequest::fontLoaded(), add logic like this:
```
if (m_font->didRefuseToLoadCustomFont() && m_document)
    m_document->addConsoleMessage(...);
```

karlcow replied:
after wrestling with the errors and the help of @brentfulgham  this morning, A new version of the PR has been pushed. 

Thanks a lot to both of you. Learning a lot in the process. 

I fixed a couple more things since that screenshot. 
![Screenshot 2023-10-11 at 8 47 33](https://github.com/WebKit/WebKit/assets/505230/6c905493-6a7f-4606-a73a-76973a630970)

cdumez commented on Source/WebCore/loader/cache/CachedFont.h:
```diff
@@ -59,6 +59,9 @@ class CachedFont : public CachedResource {
     virtual RefPtr<Font> createFont(const FontDescription&, bool syntheticBold, bool syntheticItalic, const FontCreationContext&);
 
+    bool didRefuseToLoadCustomFont() const { return m_didRefuseToLoadCustomFont; }
+    bool m_didRefuseToLoadCustomFont { false };
```
This needs to be private.

karlcow replied:
Done.  Thanks.

cdumez commented on Source/WebCore/loader/cache/CachedFontLoadRequest.h:
```diff
@@ -39,8 +40,8 @@ class FontCreationContext;
-    CachedFontLoadRequest(CachedFont& font)
-        : m_font(&font)
+    CachedFontLoadRequest(CachedFont& font, Document* document)
+        : m_font(&font), m_document(document)
```
`, m_document(document)` needs to be on the next line.

karlcow replied:
I did 

`,  m_context(context)` on the next line.

cdumez commented on Source/WebCore/loader/cache/CachedFontLoadRequest.h:
```diff
@@ -82,10 +83,16 @@ class CachedFontLoadRequest final : public FontLoadRequest, public CachedFontCli
 
     CachedResourceHandle<CachedFont> m_font;
     FontLoadRequestClient* m_fontLoadRequestClient { nullptr };
+    WeakPtr<Document, WeakPtrImplWithEventTargetData> m_document;
```
`WeakPtr<ScriptExecutionContext> m_context;`

karlcow replied:
done. thanks.

cdumez commented on Source/WebCore/css/CSSFontFaceSrcValue.cpp:
```diff
@@ -88,8 +88,10 @@ Ref<CSSFontFaceSrcResourceValue> CSSFontFaceSrcResourceValue::create(ResolvedURL
-    if (m_cachedFont)
-        return makeUnique<CachedFontLoadRequest>(*m_cachedFont);
+    if (m_cachedFont) {
+        RefPtr document = dynamicDowncast<Document>(&context);
```
No need, we can just pass the context, this way the logging will work in workers too.

karlcow replied:
hmm I passed `context` instead of `&context`. Trying to recompile with `&context`

karlcow replied:
ah no. Everything is as you asked. :) confused by my own changes.

cdumez commented on Source/WebCore/css/CSSFontFaceSrcValue.cpp:
```diff
@@ -88,8 +88,10 @@ Ref<CSSFontFaceSrcResourceValue> CSSFontFaceSrcResourceValue::create(ResolvedURL
-        return makeUnique<CachedFontLoadRequest>(*m_cachedFont);
+    if (m_cachedFont) {
+        RefPtr document = dynamicDowncast<Document>(&context);
+        return makeUnique<CachedFontLoadRequest>(*m_cachedFont, document.get());
```
`document.get()` -> `context`

karlcow replied:
Done.  Thanks.

cdumez commented on Source/WebCore/loader/cache/CachedFontLoadRequest.h:
```diff
@@ -39,8 +40,8 @@ class FontCreationContext;
 public:
-    CachedFontLoadRequest(CachedFont& font)
-        : m_font(&font)
+    CachedFontLoadRequest(CachedFont& font, Document* document)
```
`ScriptExcutionContext& context`

karlcow replied:
Done.  Thanks.

cdumez commented on Source/WebCore/loader/cache/CachedFontLoadRequest.h:
```diff
@@ -82,10 +83,16 @@ class CachedFontLoadRequest final : public FontLoadRequest, public CachedFontCli
         if (m_fontLoadRequestClient)
             m_fontLoadRequestClient->fontLoaded(*this);
+
+        if (m_font->didRefuseToLoadCustomFont() && m_document) {
```
No need for `&& m_document`

karlcow replied:
Removed it. thanks.

cdumez commented on Source/WebCore/loader/cache/CachedFontLoadRequest.h:
```diff
@@ -82,10 +83,16 @@ class CachedFontLoadRequest final : public FontLoadRequest, public CachedFontCli
+
+        if (m_font->didRefuseToLoadCustomFont() && m_document) {
+            auto message = makeString("[Lockdown Mode] This font has been blocked: ", m_font->url().string());
+            m_document->addConsoleMessage(MessageSource::Security, MessageLevel::Info, message);
```
`m_context->addConsoleMessage(...)`

karlcow replied:
done. thanks.

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3240,7 +3240,7 @@ void Document::implicitOpen()
 {
     auto* cachedFont = m_fontLoader->cachedFont(completeURL(url), isSVG, isInitiatingElementInUserAgentShadowTree, loadedFromOpaqueSource);
-    return cachedFont ? makeUnique<CachedFontLoadRequest>(*cachedFont) : nullptr;
+    return cachedFont ? makeUnique<CachedFontLoadRequest>(*cachedFont, this) : nullptr;
```
`*this`

karlcow replied:
Done.  Thanks.

cdumez commented on Source/WebCore/css/CSSFontFaceSrcValue.cpp:
```diff
@@ -88,8 +88,10 @@ Ref<CSSFontFaceSrcResourceValue> CSSFontFaceSrcResourceValue::create(ResolvedURL
-    if (m_cachedFont)
-        return makeUnique<CachedFontLoadRequest>(*m_cachedFont);
+    if (m_cachedFont) {
+        RefPtr document = dynamicDowncast<Document>(context);
```
This line is not needed.

cdumez commented on Source/WebCore/loader/cache/CachedFontLoadRequest.h:
```diff
@@ -29,6 +29,7 @@
 #include "CachedFont.h"
 #include "CachedFontClient.h"
 #include "CachedResourceHandle.h"
+#include "Document.h"
```
ScriptExecutionContext.h

cdumez reviewed (approved):
LGTM with 2 fixes.

cdumez commented on Source/WebCore/loader/cache/CachedFontLoadRequest.h:
```diff
@@ -82,10 +84,16 @@ class CachedFontLoadRequest final : public FontLoadRequest, public CachedFontCli
         if (m_fontLoadRequestClient)
             m_fontLoadRequestClient->fontLoaded(*this);
+
+        if (m_font->didRefuseToLoadCustomFont()) {
```
missing `&& m_context` check. It is a weak pointer so it can become null.

karlcow committed:
Add lockdown mode logging
https://bugs.webkit.org/show_bug.cgi?id=262518
rdar://114657783

Reviewed by Chris Dumez.

When activated Lockdown mode is reducing some of the features
accessible to a Webpage. This creates breakages which can be
sometimes a bit hard to identify. The goal of this change is
to improve the findability of the type of issues by adding logging
for engineers to understand what type of breakage is going on.

Currently we only target fonts which have not been loaded. It seems,
for now, these are the main source of breakage. We initially wanted
to log the URL of the fonts being logged in syslog, but this create
significant issues in terms of privacy, as it would share URLs which
will be private to the user context, such as intranet systems. Logging
the name of the fonts might also create issues with inline nameless
fonts and we would only know the name given to the font at the CSS
parsing time, as the blocked fonts are not being downloaded.

This also adds logging in the Web Inspector console for fonts URLs
being blocked by Lockdown mode. Being private to the user context,
this is helpful for the Web developer to determine what fonts are
missing and gives the opportunity for them to share at their will
the issue in a bug report.

We might, in the future, add more engineering logging for other
features.

If you need to learn more about lockdown mode, see
https://support.apple.com/en-us/HT212650

* Source/WebCore/loader/cache/AllowedFonts.cpp:
(WebCore::isFontBinaryAllowed):
* Source/WebCore/css/CSSFontFaceSrcValue.cpp:
(WebCore::CSSFontFaceSrcResourceValue::fontLoadRequest):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::fontLoadRequest):
* Source/WebCore/loader/cache/CachedFont.cpp:
(WebCore::CachedFont::finishLoading):
* Source/WebCore/loader/cache/CachedFont.h:
(WebCore::CachedFont::didRefuseToLoadCustomFont const):
* Source/WebCore/loader/cache/CachedFontLoadRequest.h:

Co-authored-by: Chris Dumez <cdumez@apple.com>
Co-authored-by: Brent Fulgham <bfulgham@webkit.org>
Canonical link: https://commits.webkit.org/269238@main
